### PR TITLE
Add initial redis cache support for reference datasets

### PR DESCRIPTION
**Pull request by mpiannucci**
This PR adds a redis cache over the top of the reference filesystem for kerchunk datasets, in an attempt to share a cache between multiple workers. There is no current attempt to synchronize effort (yet). Plus: 

* Correctly typed configuration via pydantic settings
* Update dockerfiles and readme to match latest configuration options
* Add `engine` option to dataset config
* Add more specific options to kerchunk dataset type
* Add `dataset_cache_timeout` to settings

The only breaking change is that the default `ROOT_PATH` is now empty and results in `/`. Any feedback is great.

### CODE REVIEW

**Comment by mpiannucci** [comment]
Testing with this tile: http://localhost:8090/datasets/wcofs/wms/?service=WMS&version=1.3.0&request=GetMap&layers=salt&crs=EPSG:3857&styles=raster/default&width=1752&height=1137&bbox=-17719381.920221075,2249529.340671755,-9552927.006119858,7549662.519481752&colorscalerange=-0.0970478355884552,34.51841354370117&time=2024-05-23T12:00:00Z&elevation=-0.0125

image

Using this branch, having turned off the in memory dataset cache we get the following:

| cache            | time |
|------------- |----- |
| redis - cold   | 6.58s  |
| redis - hot     |  0.728s | 
| redis - warm     | 4.2s | 
| file - cold      | 7.93s | 
| file - hot        | 0.926s |
| file - warm        | (NOT RELEVANT)|

File: xreds/dataset_provider.py [hunk 1]
```diff
@@ -77,8 +77,7 @@ def get_dataset(self, dataset_id: str) -> xr.Dataset:
                 logger.info(f"Applying extension {ext_name} to dataset {dataset_id}")
             ds = extension().transform_dataset(ds=ds, config=ext_config)
 
-        # TODO: For now this cache is disabled for testing with redis
-        # self.datasets[cache_key] = {"dataset": ds, "date": datetime.datetime.now()}
+        self.datasets[cache_key] = {"dataset": ds, "date": datetime.datetime.now()}
```
Comment (mpiannucci) [inline xreds/dataset_provider.py:80]: Im not sure if this should be configurable, for now this cache shares the same cache as the redis and filecache for reference datasets

File: xreds/utils.py [hunk 2]
```diff
@@ -1,74 +1,112 @@
...
+            with fsspec.open(
+                reference_url,
+                mode="rb",
+                rediscache={"redis": redis_cache, "expiry": cache_timeout},
+                s3=target_options,
+            ) as f:
+                refs = ujson.load(f)
+            fs = RedisCachingReferenceFileSystem(
+                redis=redis_cache,
+                expiry_time=180,
```
Comment (srstsavage) [inline xreds/utils.py:71]: Do we maybe want to make this configurable? `REDIS_CACHE_EXPIRY_TIME`?

Comment (mpiannucci) [reply xreds/utils.py:71]: Yup good catch, that should use the expiry config for sure

Comment (mpiannucci) [reply xreds/utils.py]: for now it will share the same value as `DATASET_CACHE_TIMEOUT`, unless you think the redis timeout should be different?

**Comment by srstsavage** [comment]
Haven't narrowed it down, but after xreds+redis runs long enough I start seeing these type of errors on requests:

```
s3fs.utils.FileExpired: [Errno 16] The remote file corresponding to filename nextgen-dmac-cloud-ingest/nos/creofs/nos.creofs.fields.best.nc.zarr and Etag "1ea3216d9
21688f125a38dbad94a3969" no longer exists.
```

Full traceback attached.

The errors persist even after restarting xreds, and only stop after clearing the redis cache.

**Comment by mpiannucci** [comment]
@srstsavage are you running with the default 10 minute cache timeout? I think this is happening because the aggregation is updated and the fileinfo is out of sync.

**Comment by srstsavage** [comment]
https://github.com/fsspec/s3fs/issues/482

**Comment by srstsavage** [comment]
https://github.com/fsspec/s3fs/commit/bf46cdc7b46cda19ede5655bf5e6a91b4b600b40

https://github.com/fsspec/s3fs/blob/main/s3fs/utils.py#L93-L97

```
class FileExpired(IOError):
    """
    Is raised, when the file content has been changed from a different process after
    opening the file. Reading the file would lead to invalid or inconsistent output.
    This can also be triggered by outdated file-information inside the directory cache.
    In this case ``S3FileSystem.invalidate_cache`` can be used to force an update of
    the file-information when opening the file.
    """    
```

**Comment by mpiannucci** [comment]
Added in update to `redis_fsspec_cache` from @srstsavage and will test a few times over the next day or so to make sure the invalidation error doesnt happen anymore

**Comment by mpiannucci** [comment]
@nguyandy just a ping on the changes to the docker processes here

**Comment by mpiannucci** [comment]
I think this is good now, and it is opt in and improves the documentation so I am going to merge

**Pull request merged**